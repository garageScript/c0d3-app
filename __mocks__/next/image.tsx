import React from 'react'
import type { ImageProps } from 'next/image'
import _ from 'lodash'

const Image = (props: ImageProps) => {
  /*not all Image properties are allowed in <img/> tag
  ObjectFit causes warnings, but otherwise lessons/error images won't keep their dimensions 
  see https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit*/
  const copy = _.omit(props, 'objectFit')
  return <img {...copy} />
}

export default Image
