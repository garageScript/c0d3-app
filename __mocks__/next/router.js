const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
