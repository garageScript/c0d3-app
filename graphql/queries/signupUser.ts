import { gql } from '@apollo/client'

const SIGNUP_USER = gql`
  mutation signup(
    $firstName: String!
    $lastName: String!
    $email: String!
    $username: String!
  ) {
    signup(
      firstName: $firstName
      lastName: $lastName
      email: $email
      username: $username
    ) {
      success
      username
      error
    }
  }
`

export default SIGNUP_USER
