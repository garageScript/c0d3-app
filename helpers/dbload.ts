import { Options, Sequelize } from 'sequelize'
import { Lesson, LessonTypes } from './models/Lesson'
import { User, UserTypes } from './models/User'
import { UserLessonTypes, UserLesson } from './models/UserLesson'
import { SubmissionTypes, Submission } from './models/Submission'
import { ChallengeTypes, Challenge } from './models/Challenge'
import { AlertTypes, Alert } from './models/Alert'
import { StarTypes, Star } from './models/Star'

const options: Options = {
  host: process.env.DB_HOST || 'city',
  logging: false,
  dialect: 'postgres',
  pool: {
    max: 5,
    min: 0,
    acquire: 30000,
    idle: 10000
  }
}

if (process.env.NODE_ENV != 'production') {
  options.port = ((process.env.DB_PORT as unknown) as number) || 5432
}
const sequelize = new Sequelize(
  process.env.DB_NAME || 'you',
  process.env.DB_USER || 'failed',
  process.env.DB_PW || 'this',
  options
)

Alert.init(AlertTypes, {
  tableName: 'alerts',
  sequelize
})

Lesson.init(LessonTypes, {
  tableName: 'lessons',
  sequelize
})

User.init(UserTypes, {
  tableName: 'users',
  sequelize
})

UserLesson.init(UserLessonTypes, {
  tableName: 'userLessons',
  sequelize
})

Submission.init(SubmissionTypes, {
  tableName: 'submissions',
  sequelize
})

Challenge.init(ChallengeTypes, {
  tableName: 'challenges',
  sequelize
})

Star.init(StarTypes, {
  tableName: 'stars',
  sequelize
})

Lesson.hasMany(Challenge, {
  as: 'challenges', // Defaults to Challenges
  foreignKey: 'lessonId'
})
Lesson.belongsToMany(User, {
  as: 'lesson',
  foreignKey: 'lessonId', // Defaults to LessonId
  through: { model: UserLesson }
})

Submission.belongsTo(User, { as: 'user' })
Submission.belongsTo(User, { as: 'reviewer' })
Submission.belongsTo(Challenge, { as: 'challenge' })
Submission.belongsTo(Lesson, { as: 'lesson' })

Star.belongsTo(User, { as: 'student' })
Star.belongsTo(User, { as: 'mentor' })
Star.belongsTo(Lesson, { as: 'lesson' })

User.belongsToMany(Lesson, {
  foreignKey: 'userId', // Defaults to UserId
  through: { model: UserLesson }
})

User.hasMany(Submission, {
  foreignKey: 'userId'
})

User.hasMany(UserLesson, {
  foreignKey: 'userId'
})

UserLesson.belongsTo(User, {
  foreignKey: 'userId',
  as: 'user'
})

sequelize.sync({ alter: !!process.env.ALTER_DB })

export default {
  Lesson,
  Challenge,
  Submission,
  User,
  UserLesson,
  Alert,
  Star,
  sequelize
}

export { User, UserLesson, options }
