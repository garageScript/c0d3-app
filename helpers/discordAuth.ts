import prisma from '../prisma'
import { URLSearchParams } from 'url'
import { User } from '.prisma/client'
import fetch from 'node-fetch'

const discordAPI = 'https://discordapp.com/api'
export const client_id = process.env.DISCORD_KEY
export const client_secret = process.env.DISCORD_SECRET
export const redirect_uri = process.env.DISCORD_REDIRECT_URI // {baseurl}/api/auth/callback/discord

type AccessTokenResponse = {
  access_token: string
  token_type: string
  expires_in: number
  refresh_token: string
  scope: string
}

type UserInfoResponse = {
  id: string
  username: string
  avatar: string
  discriminator: string
  public_flags: number
  locale: string
  mfa_enabled: boolean
  email: string
  verified: boolean
}

export type DiscordUserInfo = {
  userId: string
  username: string
  avatarUrl: string
  refreshToken: string
}

export const updateRefreshandAccessTokens = (
  userId: number,
  discordId: string,
  refreshToken: string,
  accessToken: string,
  accessTokenExpiresAt: Date
): Promise<User> => {
  return prisma.user.update({
    where: {
      id: userId
    },
    data: {
      discordRefreshToken: refreshToken,
      discordAccessToken: accessToken,
      discordAccessTokenExpires: accessTokenExpiresAt,
      discordId
    }
  })
}

const getTokenFromRefreshToken = (
  refresh_token: string
): Promise<AccessTokenResponse> => {
  return fetch(`${discordAPI}/oauth2/token`, {
    method: 'POST',
    headers: {
      'content-type': 'application/x-www-form-urlencoded;charset=utf-8'
    },
    body: new URLSearchParams({
      client_id,
      client_secret,
      grant_type: 'refresh_token',
      refresh_token
    })
  }).then(r => r.json())
}

const getUserInfoFromAccessToken = (
  accessToken: string
): Promise<UserInfoResponse> => {
  return fetch(`${discordAPI}/users/@me`, {
    method: 'GET',
    headers: {
      'content-type': 'application/x-www-form-urlencoded;charset=utf-8',
      Authorization: `Bearer ${accessToken}`
    }
  }).then(r => r.json())
}

export const getDiscordUserInfo = async (
  user: User
): Promise<DiscordUserInfo> => {
  try {
    const {
      discordRefreshToken,
      discordAccessToken,
      discordAccessTokenExpires
    } = user

    let userId = '',
      username = '',
      avatarUrl = '',
      refreshToken = discordRefreshToken || '',
      accessToken = discordAccessToken || ''

    if (!discordRefreshToken)
      return {
        userId,
        username,
        avatarUrl,
        refreshToken
      }

    if (!discordAccessToken || Number(discordAccessTokenExpires) < Date.now()) {
      const tokenResponse = await getTokenFromRefreshToken(refreshToken)
      refreshToken = tokenResponse.refresh_token || ''
      accessToken = tokenResponse.access_token || ''
      const accessTokenExpiresAt = new Date(
        Date.now() + (tokenResponse.expires_in * 1000 || 0)
      )

      // if updatedRefreshToken is undefined, empty string is stored in db to remove invalid tokens
      await updateRefreshandAccessTokens(
        user.id,
        user.discordId!,
        refreshToken,
        accessToken,
        accessTokenExpiresAt
      )
    }

    if (!accessToken)
      throw new Error(`refresh token invalid for userId ${user.id}`)

    const userInfo = await getUserInfoFromAccessToken(accessToken)

    userId = userInfo.id
    username = `${userInfo.username}#${userInfo.discriminator}`
    avatarUrl = `https://cdn.discordapp.com/avatars/${userInfo.id}/${userInfo.avatar}.png`

    if (!userId) throw new Error(`access token invalid for userId ${user.id}`)

    return {
      userId,
      username,
      avatarUrl,
      refreshToken
    }
  } catch (error) {
    return {
      userId: '',
      username: '',
      avatarUrl: '',
      refreshToken: ''
    }
  }
}
