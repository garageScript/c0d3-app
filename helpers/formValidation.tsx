import * as Yup from 'yup'

const TEXT_MIN = 2
const TEXT_MAX = 64
const PASSWORD_MIN = 6
const REGEX_ALPHANUMERICS_AND_SPACE = /^[a-zA-Z0-9_\s]*$/

const alertValidation = Yup.object({
  text: Yup.string().required('Required').strict(true),
  type: Yup.string().required('Required').strict(true),
  url: Yup.string(),
  urlCaption: Yup.string()
})

const lessonSchema = Yup.object({
  title: Yup.string().required('Required').strict(true),
  description: Yup.string().required('Required').strict(true),
  order: Yup.number()
    .required('Required')
    .typeError('Numbers only')
    .min(0, 'Positive numbers only')
    .strict(true)
})

const signupValidation = Yup.object({
  email: Yup.string().email('Invalid email address').required('Required'),
  username: Yup.string()
    .strict(true)
    .lowercase('Must be a lowercase string')
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required'),
  firstName: Yup.string()
    .strict(true)
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required'),
  lastName: Yup.string()
    .strict(true)
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required')
})

const loginValidation = Yup.object({
  username: Yup.string()
    .strict(true)
    .lowercase('Must be a lowercase string')
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required'),
  password: Yup.string()
    .min(PASSWORD_MIN, `Must be at least ${PASSWORD_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .required('Required')
})

const passwordValidation = Yup.object({
  password: Yup.string()
    .min(PASSWORD_MIN, `Must be at least ${PASSWORD_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .required('Required')
})

const confirmPasswordValidation = Yup.object({
  password: Yup.string()
    .min(PASSWORD_MIN, `Must be at least ${PASSWORD_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .required('Required'),
  confirmPassword: Yup.string()
    .oneOf([Yup.ref('password'), undefined], 'Password must match')
    .required('Required')
})

const resetPasswordValidation = Yup.object({
  userOrEmail: Yup.string()
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .required('Required')
})

export {
  alertValidation,
  lessonSchema,
  signupValidation,
  loginValidation,
  passwordValidation,
  confirmPasswordValidation,
  resetPasswordValidation
}
