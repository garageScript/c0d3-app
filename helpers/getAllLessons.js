function importAll(r) {
  return r.keys().map(fileName => ({
    link: fileName.substr(1).replace(/\/index\.mdx$/, ''),
    module: r(fileName)
  }))
}

export const lessons = importAll(
  require.context('../pages/curriculum/lessons', true, /\.mdx$/)
)
