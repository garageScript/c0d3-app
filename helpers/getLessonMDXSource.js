import path from 'path'
import { LESSONS_PATH } from '../lib/CONSTANTS'
import { serialize } from 'next-mdx-remote/serialize'
import matter from 'gray-matter'
const slug2 = require('remark-slug')
const toc = require('remark-toc')
const gfm = require('remark-gfm')
export async function getLessonMDXSource(slug) {
  const fs = require('fs')
  const lessonFilePath = path.join(LESSONS_PATH, slug, 'lesson.mdx')
  console.log(lessonFilePath)
  const source = fs.readFileSync(lessonFilePath)

  const { content, data } = matter(source)

  const mdxSource = await serialize(content, {
    // Optionally pass remark/rehype plugins
    mdxOptions: {
      remarkPlugins: [slug2, [toc, { maxDepth: 2 }], gfm],
      rehypePlugins: []
    },
    scope: data
  })

  // const source = 'Some **mdx** text, with a component'
  // const mdxSource = await serialize(source)
  console.log('thesourcest', mdxSource.compiledSource)
  return mdxSource
}
