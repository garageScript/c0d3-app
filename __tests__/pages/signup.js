jest.mock('next/router')
import React from 'react'
import { render, fireEvent, wait, act } from '@testing-library/react'
import Signup from '../../pages/signup'
import * as signupHelper from '../../helpers/signupUser'
import { useRouter } from 'next/router'

useRouter.mockReturnValue({
  route: null
})

describe('Signup Page', () => {
  const fillOutSignupForm = async getByTestId => {
    const emailField = getByTestId('email')
    const usernameField = getByTestId('username')
    const passwordField = getByTestId('password')
    const firstNameField = getByTestId('firstName')
    const lastNameField = getByTestId('lastName')
    await wait(
      fireEvent.change(emailField, {
        target: {
          value: 'email@domain.com'
        }
      }),
      fireEvent.change(usernameField, {
        target: {
          value: 'user name'
        }
      }),
      fireEvent.change(passwordField, {
        target: {
          value: 'password123'
        }
      }),
      fireEvent.change(firstNameField, {
        target: {
          value: 'user'
        }
      }),
      fireEvent.change(lastNameField, {
        target: {
          value: 'name'
        }
      })
    )
  }

  test('Should render without crashing', () => {
    const { getByTestId } = render(<Signup />)
    getByTestId('email')
    getByTestId('username')
    getByTestId('password')
    getByTestId('firstName')
    getByTestId('lastName')
  })

  test('Should not submit values if form is empty', async () => {
    signupHelper.signupUser = jest.fn()
    const { getByTestId } = render(<Signup />)
    const submitButton = getByTestId('submit')
    await wait(() => {
      act(() => {
        fireEvent.click(submitButton)
      }),
        expect(signupHelper.signupUser).not.toBeCalled()
    })
  })

  test('Should submit signup form values and render success component', async () => {
    signupHelper.signupUser = jest
      .fn()
      .mockReturnValue(Promise.resolve({ success: true }))
    const { getByTestId } = render(<Signup />)
    const submitButton = getByTestId('submit')
    fillOutSignupForm(getByTestId)

    await wait(() => {
      act(() => {
        fireEvent.click(submitButton)
      }),
        expect(signupHelper.signupUser).toHaveBeenCalledTimes(1)
    })
  })

  test('Should submit signup form values and display error component upon failure of email', async () => {
    signupHelper.signupUser = jest.fn().mockReturnValue(
      Promise.resolve({
        success: false,
        errorMessage: {
          confirmEmail: ['This email has been taken']
        }
      })
    )
    const { container, getByTestId } = render(<Signup />)
    const submitButton = getByTestId('submit')
    fillOutSignupForm(getByTestId)

    await wait(() => {
      act(() => {
        fireEvent.click(submitButton)
      }),
        expect(signupHelper.signupUser).toHaveBeenCalledTimes(1),
        expect(container).toMatchSnapshot()
    })
  })

  test('Should submit signup form values and display error component upon failure of username', async () => {
    signupHelper.signupUser = jest.fn().mockReturnValue(
      Promise.resolve({
        success: false,
        errorMessage: {
          userName: ['Username not available']
        }
      })
    )
    const { container, getByTestId } = render(<Signup />)
    const submitButton = getByTestId('submit')
    fillOutSignupForm(getByTestId)

    await wait(() => {
      act(() => {
        fireEvent.click(submitButton)
      }),
        expect(signupHelper.signupUser).toHaveBeenCalledTimes(1),
        expect(container).toMatchSnapshot()
    })
  })

  test('Should submit signup form values and display error component upon failure of both username and email', async () => {
    signupHelper.signupUser = jest.fn().mockReturnValue(
      Promise.resolve({
        success: false,
        errorMessage: {
          confirmEmail: ['This email has been taken'],
          userName: ['Username not available']
        }
      })
    )
    const { container, getByTestId } = render(<Signup />)
    const submitButton = getByTestId('submit')
    fillOutSignupForm(getByTestId)

    await wait(() => {
      act(() => {
        fireEvent.click(submitButton)
      }),
        expect(signupHelper.signupUser).toHaveBeenCalledTimes(1),
        expect(container).toMatchSnapshot()
    })
  })
})
