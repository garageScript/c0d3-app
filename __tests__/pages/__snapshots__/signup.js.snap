// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Page Should submit signup form values and display error component upon failure of both username and email 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/home"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/"
            >
              Home 
              <span
                class="sr-only"
              >
                (current)
              </span>
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Create Account
          </h1>
          <p
            class="card-text"
          />
          <div
            class="bg-light m-auto px-5 border-0'"
          >
            <h5
              class="text-danger"
            >
              An account has already been registered with this email
            </h5>
            <h5
              class="text-danger"
            >
              Username is already taken
            </h5>
          </div>
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="form-group "
            >
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="email"
                name="email"
                placeholder="Email address"
                type="email"
                value="email@domain.com"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="username"
                name="username"
                placeholder="Username"
                value="user name"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="password"
                name="password"
                placeholder="Password"
                type="password"
                value="password123"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="firstName"
                name="firstName"
                placeholder="First name"
                value="user"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="lastName"
                name="lastName"
                placeholder="Last name"
                value="name"
              />
              <button
                class="btn btn-primary btn-lg btn-block mb-3"
                data-testid="submit"
                type="submit"
              >
                Create Account
              </button>
            </div>
          </form>
          <p
            class="text-black-50"
          >
            Already have an account?
             
            <a
              class="text-primary"
              href="/login"
            >
              Login
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;

exports[`Signup Page Should submit signup form values and display error component upon failure of email 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/home"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/"
            >
              Home 
              <span
                class="sr-only"
              >
                (current)
              </span>
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Create Account
          </h1>
          <p
            class="card-text"
          />
          <div
            class="bg-light m-auto px-5 border-0'"
          >
            <h5
              class="text-danger"
            >
              An account has already been registered with this email
            </h5>
            
          </div>
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="form-group "
            >
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="email"
                name="email"
                placeholder="Email address"
                type="email"
                value="email@domain.com"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="username"
                name="username"
                placeholder="Username"
                value="user name"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="password"
                name="password"
                placeholder="Password"
                type="password"
                value="password123"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="firstName"
                name="firstName"
                placeholder="First name"
                value="user"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="lastName"
                name="lastName"
                placeholder="Last name"
                value="name"
              />
              <button
                class="btn btn-primary btn-lg btn-block mb-3"
                data-testid="submit"
                type="submit"
              >
                Create Account
              </button>
            </div>
          </form>
          <p
            class="text-black-50"
          >
            Already have an account?
             
            <a
              class="text-primary"
              href="/login"
            >
              Login
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;

exports[`Signup Page Should submit signup form values and display error component upon failure of username 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/home"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/"
            >
              Home 
              <span
                class="sr-only"
              >
                (current)
              </span>
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Create Account
          </h1>
          <p
            class="card-text"
          />
          <div
            class="bg-light m-auto px-5 border-0'"
          >
            
            <h5
              class="text-danger"
            >
              Username is already taken
            </h5>
          </div>
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="form-group "
            >
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="email"
                name="email"
                placeholder="Email address"
                type="email"
                value="email@domain.com"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="username"
                name="username"
                placeholder="Username"
                value="user name"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="password"
                name="password"
                placeholder="Password"
                type="password"
                value="password123"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="firstName"
                name="firstName"
                placeholder="First name"
                value="user"
              />
              <input
                class="form-control form-control-lg font-weight-light mb-3 "
                data-testid="lastName"
                name="lastName"
                placeholder="Last name"
                value="name"
              />
              <button
                class="btn btn-primary btn-lg btn-block mb-3"
                data-testid="submit"
                type="submit"
              >
                Create Account
              </button>
            </div>
          </form>
          <p
            class="text-black-50"
          >
            Already have an account?
             
            <a
              class="text-primary"
              href="/login"
            >
              Login
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;
